pulsar-io/nifi/src/main/java/org/apache/pulsar/io/nifi/NiFiSink.java

```diff
@@ -0,0 +1,156 @@
+
+    private int requestBatchCount;
+    private long waitTimeMs;
+    private List<Record<NiFiDataPacket>>  currentList;
+    private ScheduledExecutorService flushExecutor;
+
+    @Override
+    public void open(Map<String, Object> config, SinkContext sinkContext) throws Exception {
+        niFiConfig = NiFiConfig.load(config);
+        Preconditions.checkNotNull(niFiConfig.getUrl(), "url property not set.");
+        Preconditions.checkNotNull(niFiConfig.getPortName(), "portName property not set.");
+        Preconditions.checkArgument(niFiConfig.getRequestBatchCount() > 0,
+                "requestBatchCount must be a positive integer.");
+
+        clientConfig = new SiteToSiteClient.Builder()
+                .url(niFiConfig.getUrl())
+                .portName(niFiConfig.getPortName())
+                .requestBatchCount(niFiConfig.getRequestBatchCount())
+                .buildConfig();
+        client = new SiteToSiteClient.Builder().fromConfig(clientConfig).build();
+
+        requestBatchCount = niFiConfig.getRequestBatchCount();
+        waitTimeMs = niFiConfig.getWaitTimeMs();
+        currentList= Lists.newArrayList();
+
+        flushExecutor = Executors.newScheduledThreadPool(1);
+        flushExecutor.scheduleAtFixedRate(() -> flush(), waitTimeMs, waitTimeMs, TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (null != client) {
+            client.close();
+        }
+
+        if (null != flushExecutor) {
+            flushExecutor.shutdown();
+        }
+
+    }
+
+    @Override
+    public void write(Record<NiFiDataPacket> record) {
+        int number;
+        synchronized (this) {
+            if (null != record) {
+                currentList.add(record);
+            }
+            number = currentList.size();
+        }
+
+        if (number == requestBatchCount) {
+            flushExecutor.submit(() -> flush());
+        }
+    }
+
+    private void flush() {
+        List<Record<NiFiDataPacket>>  toFlushList;
+        synchronized (this) {
+            if (currentList.isEmpty()) {
+                return;
+            }
+            toFlushList = currentList;
+            currentList = Lists.newArrayList();
+        }
+
+        Transaction transaction = null;
+        try {
+            transaction = client.createTransaction(TransferDirection.SEND);
+        } catch (IOException ioe) {
+            log.warn("Created NiFi transaction Failed", ioe);
+        }
+
+        if (null != transaction) {
+            if (CollectionUtils.isNotEmpty(toFlushList)) {
+                for (Record<NiFiDataPacket> record : toFlushList) {
+                    NiFiDataPacket niFiDataPacket = record.getValue();
+                    try {
+                        transaction.send(niFiDataPacket.getContent(), niFiDataPacket.getAttributes());
+                    } catch (IOException ioe) {
+                        log.warn("Failed to send data to NiFi", ioe);
```

Comment:
if you fail to send the record, don't you need to error out the record?

Reply:
> if you fail to send the record, don't you need to error out the record?

ok